# lumpy_filt.pl

This program filters the output from LUMPY.

# LUMPY output format

The program accepts as input a .vcf file generated by running lumpyexpress.

It assumes that data have been processed as follows: 

### Run lumpyexpress [https://github.com/arq5x/lumpy-sv]

```{bash}
lumpyexpress \
	-B tumour.bam,normal.bam \
	-S tumour.split.sort.bam,normal.split.sort.bam \
    -D tumour.discordant.sort.bam,normal.discordant.sort.bam \
    -o tumour.normal.vcf
```

### Genotype output from lumpyexpress with svtools genotype [https://github.com/hall-lab/svtools]

```{bash}
svtools genotype \
	-i tumour.normal.vcf \
	-o tumour.normal.gt.vcf \
	-B tumour.bam,normal.bam
```

### Data should look something like this:

```
##CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  TUMOUR   NORMAL
2L      605109  1       N       <DUP>   124.73  .       SVTYPE=DUP;SVLEN=912;END=606021;STRANDS=-+:25;CIPOS=-10,9;CIEND=-2,1;CIPOS95=0,0;CIEND95=0,0;SU=25;PE=9;SR=16   GT:SU:PE:SR:GQ:SQ:GL:DP:RO:AO:QR:QA:RS:AS:ASC:RP:AP:AB  0/1:12:4:8:24:82.12:-10,-1,-4:44:33:10:32:9:22:5:0:10:4:0.22    0/1:13:5:8:42:42.61:-8,-4,-9:67:56:10:55:9:34:4:0:21:5:0.14
2L      18885   2       N       <DEL>   50.75   .       SVTYPE=DEL;SVLEN=-2794;END=21679;STRANDS=+-:7;IMPRECISE;CIPOS=-10,458;CIEND=-361,9;CIPOS95=-3,60;CIEND95=-54,3;SU=7;PE=7;SR=0   GT:SU:PE:SR:GQ:SQ:GL:DP:RO:AO:QR:QA:RS:AS:ASC:RP:AP:AB  0/0:2:2:0:3:2.98:-2,-2,-8:11:9:1:9:1:0:0:0:9:1:0.1      0/1:5:5:0:47:47.78:-10,-5,-24:35:29:5:29:5:0:0:0:29:5:0.15
2L      25371   3       N       <DUP>   50.22   .       SVTYPE=DUP;SVLEN=396;END=25767;STRANDS=-+:5;IMPRECISE;CIPOS=-10,9;CIEND=-10,9;CIPOS95=-1,0;CIEND95=-1,1;SU=5;PE=3;SR=2  GT:SU:PE:SR:GQ:SQ:GL:DP:RO:AO:QR:QA:RS:AS:ASC:RP:AP:AB  0/0:0:0:0:24:0.02:-1,-3,-6:25:23:1:23:1:0:0:0:23:1:0.042        0/1:5:3:2:19:50.21:-6,-1,-3:31:24:6:24:6:0:0:0:24:6:0.2
2L      47519   4       N       <DEL>   2629.28 .       SVTYPE=DEL;SVLEN=-5040;END=52559;STRANDS=+-:84;CIPOS=0,0;CIEND=0,0;CIPOS95=0,0;CIEND95=0,0;SU=84;PE=38;SR=46    GT:SU:PE:SR:GQ:SQ:GL:DP:RO:AO:QR:QA:RS:AS:ASC:RP:AP:AB  1/1:38:14:24:104:1211.24:-123,-12,-2:42:0:42:0:41:0:23:4:0:13:1 1/1:46:24:22:122:1418.04:-144,-14,-2:49:0:49:0:48:0:21:3:0:23:1
2L      134916  5       N       <DEL>   134.76  .       SVTYPE=DEL;SVLEN=-32;END=134948;STRANDS=+-:6;CIPOS=-8,6;CIEND=-7,6;CIPOS95=0,0;CIEND95=0,0;SU=6;PE=1;SR=5       GT:SU:PE:SR:GQ:SQ:GL:DP:RO:AO:QR:QA:RS:AS:ASC:RP:AP:AB  0/1:4:1:3:41:59.93:-7,-1,-5:11:7:3:7:3:7:0:2:0:0:0.3    0/1:2:0:2:74:74.83:-10,-3,-16:26:20:5:20:5:20:0:5:0:0:0.2
2L      245133  6       N       <DEL>   186.00  .       SVTYPE=DEL;SVLEN=-35;END=245168;STRANDS=+-:8;CIPOS=-10,9;CIEND=-10,9;CIPOS95=0,0;CIEND95=0,0;SU=8;PE=0;SR=8     GT:SU:PE:SR:GQ:SQ:GL:DP:RO:AO:QR:QA:RS:AS:ASC:RP:AP:AB  0/1:3:0:3:12:12.17:-4,-3,-12:17:14:2:14:2:14:0:2:0:0:0.12       0/1:5:0:5:77:173.83:-19,-1,-9:23:14:8:14:8:14:0:8:0:0:0.36
2L      263351  7       N       <DEL>   1270.32 .       SVTYPE=DEL;SVLEN=-94;END=263445;STRANDS=+-:29;CIPOS=-2,0;CIEND=-1,0;CIPOS95=0,0;CIEND95=0,0;SU=29;PE=1;SR=28    GT:SU:PE:SR:GQ:SQ:GL:DP:RO:AO:QR:QA:RS:AS:ASC:RP:AP:AB  1/1:13:0:13:51:590.85:-60,-6,-1:20:0:20:0:20:0:12:7:0:0:1       1/1:16:1:15:58:679.48:-69,-7,-1:23:0:23:0:23:0:14:8:0:0:1
2L      297740  8       N       <DEL>   30.07   .       SVTYPE=DEL;SVLEN=-44;END=297784;STRANDS=+-:5;CIPOS=-10,9;CIEND=-10,9;CIPOS95=0,0;CIEND95=0,0;SU=5;PE=0;SR=5     GT:SU:PE:SR:GQ:SQ:GL:DP:RO:AO:QR:QA:RS:AS:ASC:RP:AP:AB  0/0:1:0:1:15:0.13:-2,-3,-13:16:14:1:14:1:14:0:1:0:0:0.067       0/1:4:0:4:29:29.94:-4,-1,-6:11:8:2:8:2:8:0:2:0:0:0.2
```


Run as `perl lumpy_filt.pl -h` to see usage statement

********** lumpy_filter ***********
Usage: lumpy_filt.pl [options]
  --vcf = lumpy output (see README.txt for instructions on generating this)
  --output_dir = specify name of output directory to write to
  --help
Nick Riddiford 2017